File: C_Sharp/Framework/ForeignKeyMappingManager.cs
using System;
using System.Collections.Generic;
using Framework.Domain;

namespace Framework
{
    public interface IForeignKeyMappingManager
    {
        void RegisterForeignKeyMapping<TParentEntity, TChildEntity>(
            IForeignKeyMapping<TParentEntity, TChildEntity> mapping)
            where TParentEntity : IDomainObject
            where TChildEntity : IDomainObject;

        IList<TChildEntity> GetForeignKeyValues<TParentEntity, TChildEntity>(TParentEntity entity)
            where TParentEntity : IDomainObject
            where TChildEntity : IDomainObject;
    }

    public class ForeignKeyMappingManager : IForeignKeyMappingManager
    {
        static IForeignKeyMappingManager s_instance = new ForeignKeyMappingManager();

        IDictionary<string, IDictionary<string, object>> _foreignKeMapping = new Dictionary<string, IDictionary<string, object>>();

        private ForeignKeyMappingManager()
        {
        }

        public static IForeignKeyMappingManager GetInstance()
        {
            return s_instance;
        }

        public void RegisterForeignKeyMapping<TParentEntity, TChildEntity>(
            IForeignKeyMapping<TParentEntity, TChildEntity> mapping)
            where TParentEntity : IDomainObject
            where TChildEntity : IDomainObject
        {
            string parentName = typeof(TParentEntity).FullName;

            if (!_foreignKeMapping.ContainsKey(parentName))
                _foreignKeMapping.Add(parentName, new Dictionary<string, object>());

            IDictionary<string, object> maps = _foreignKeMapping[parentName];
            string childName = typeof(TChildEntity).FullName;

            if (!maps.ContainsKey(childName))
                maps.Add(childName, mapping);
        }

        public IList<TChildEntity> GetForeignKeyValues<TParentEntity, TChildEntity>(TParentEntity entity)
            where TParentEntity : IDomainObject
            where TChildEntity : IDomainObject
        {
            string parentName = typeof(TParentEntity).FullName;

            if (!_foreignKeMapping.ContainsKey(parentName))
                throw new ArgumentException(string.Format("Foreign key mapping associated to parent type '{0}' does not exists", parentName));

            IDictionary<string, object> maps = _foreignKeMapping[parentName];
            string childName = typeof(TChildEntity).FullName;

            if (!maps.ContainsKey(childName))
                throw new ArgumentException(string.Format("Foreign key mapping associated to child type '{0}' does not exists", childName));

            IForeignKeyMapping<TParentEntity, TChildEntity> fkMapping = (IForeignKeyMapping<TParentEntity, TChildEntity>)maps[childName];

            return fkMapping.GetChildEntities(entity);
        }
    }
}

File: Android/framework/src/main/java/com/ForeignKeyMappingManager.java
package com;

import com.Interfaces.IForeignKeyMappingManager;
import com.domain.DomainObjectInterfaces.IDomainObject;
import com.domain.ForeignKeyMappingInterfaces.IForeignKeyMapping;

import java.rmi.NoSuchObjectException;
import java.util.HashMap;
import java.util.List;

/**
 * Created by aiko on 4/1/17.
 */

public class ForeignKeyMappingManager implements IForeignKeyMappingManager {
    static IForeignKeyMappingManager s_instance = new ForeignKeyMappingManager();

    HashMap<String, HashMap<String, Object>> _foreignKeMapping = new HashMap<String, HashMap<String, Object>>();

    private ForeignKeyMappingManager(){}

    public static IForeignKeyMappingManager GetInstance()
    {
        return s_instance;
    }

    public <TParentEntity extends IDomainObject, TChildEntity extends IDomainObject>
    void RegisterForeignKeyMapping(Class<TParentEntity> parentClass, Class<TChildEntity> childClass, IForeignKeyMapping<TParentEntity, TChildEntity> mapping)
    {
        String parentName = parentClass.getName();

        if (!_foreignKeMapping.containsKey(parentName))
            _foreignKeMapping.put(parentName, new HashMap<String, Object>());

        HashMap<String, Object> maps = _foreignKeMapping.get(parentName);
        String childName = childClass.getName();

        if (!maps.containsKey(childName))
            maps.put(childName, mapping);
    }

    public <TParentEntity extends IDomainObject, TChildEntity extends IDomainObject>
    List<TChildEntity> GetForeignKeyValues(Class<TParentEntity> parentClass, Class<TChildEntity> childClass, TParentEntity entity)
            throws NoSuchObjectException
    {
        String parentName = parentClass.getName();

        if (!_foreignKeMapping.containsKey(parentName))
            throw new NoSuchObjectException(String.format("Foreign key mapping associated to parent type '%s' does not exists", parentName));

        HashMap<String, Object> maps = _foreignKeMapping.get(parentName);
        String childName = childClass.getName();

        if (!maps.containsKey(childName))
            throw new NoSuchObjectException(String.format("Foreign key mapping associated to child type '%s' does not exists", childName));

        IForeignKeyMapping<TParentEntity, TChildEntity> fkMapping = (IForeignKeyMapping<TParentEntity, TChildEntity>)maps.get(childName);

        return fkMapping.GetChildEntities(entity);
    }
}
